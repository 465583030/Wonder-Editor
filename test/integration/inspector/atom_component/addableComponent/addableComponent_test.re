open Wonder_jest;

open Expect;

open Expect.Operators;

open Sinon;

let _ =
  describe("AddableComponent", () => {
    let sandbox = getSandboxDefaultVal();
    beforeEach(() => {
      sandbox := createSandbox();
      MainEditorSceneTool.initState(~sandbox, ());
    });
    afterEach(() => restoreSandbox(refJsObjToSandbox(sandbox^)));

    describe("test show component", () => {
      test("if hasn't currentSceneTreeNode, show nothing", () => {
        MainEditorSceneTool.createDefaultScene(
          sandbox,
          MainEditorSceneTool.setFirstBoxTobeCurrentSceneTreeNode,
        );

        BuildComponentTool.buildInspectorComponent(
          TestTool.buildEmptyAppState(),
          InspectorTool.buildFakeAllShowComponentConfig(),
        )
        |> ReactTestTool.createSnapshotAndMatch;
      });

      describe("else", () => {
        test(
          "if currentSceneTreeNode is camera, should show transform and basicCameraView and perspectiveCameraProjection",
          () => {
            MainEditorSceneTool.createDefaultScene(
              sandbox,
              MainEditorSceneTool.setCameraTobeCurrentSceneTreeNode,
            );
            BuildComponentTool.buildInspectorComponent(
              TestTool.buildEmptyAppState(),
              InspectorTool.buildFakeAllShowComponentConfig(),
            )
            |> ReactTestTool.createSnapshotAndMatch;
          },
        );

        test(
          "else if currentSceneTreeNode is box, should show transform and material",
          () => {
          MainEditorSceneTool.createDefaultScene(
            sandbox,
            MainEditorSceneTool.setFirstBoxTobeCurrentSceneTreeNode,
          );
          BuildComponentTool.buildInspectorComponent(
            TestTool.buildEmptyAppState(),
            InspectorTool.buildFakeAllShowComponentConfig(),
          )
          |> ReactTestTool.createSnapshotAndMatch;
        });
      });
    });
    describe("test gameObject add component workflow", () => {
      beforeEach(() => {
        MainEditorSceneTool.createDefaultScene(
          sandbox,
          MainEditorSceneTool.setFirstBoxTobeCurrentSceneTreeNode,
        );

        CurrentSelectSourceEditorService.setCurrentSelectSource(
          EditorType.SceneTree,
        )
        |> StateLogicService.getAndSetEditorState;
      });
      test("click the add component button, show addableComponent list", () => {
        let component =
          BuildComponentTool.buildInspectorComponent(
            TestTool.buildEmptyAppState(),
            InspectorTool.buildFakeAllShowComponentConfig(),
          );
        let boxComponentCount = ComponentDomTool.getBoxComponentCount();

        BaseEventTool.triggerComponentEvent(
          component,
          OperateComponentEventTool.triggerClickShowComponentList(
            boxComponentCount,
          ),
        );
        component |> ReactTestTool.createSnapshotAndMatch;
      });
      /* test("click sourceInstance component, add to inspector", () => {
           let component =
             BuildComponentTool.buildInspectorComponent(
               TestTool.buildEmptyAppState(),
               InspectorTool.buildFakeAllShowComponentConfig(),
             );
           let boxComponentCount = ComponentDomTool.getBoxComponentCount();

           BaseEventTool.triggerComponentEvent(
             component,
             OperateComponentEventTool.triggerClickShowComponentList,
           );
           BaseEventTool.triggerComponentEvent(
             component,
             OperateComponentEventTool.triggerClickAddSourceInstanceEvent,
           );

           BuildComponentTool.buildInspectorComponent(
             TestTool.buildEmptyAppState(),
             InspectorTool.buildFakeAllShowComponentConfig(),
           )
           |> ReactTestTool.createSnapshotAndMatch;
         }); */
      test("click light component, add to inspector", () => {
        let boxComponentCount = ComponentDomTool.getBoxComponentCount();
        let renderingCategoryDomIndex =
          ComponentDomTool.getRenderingCategoryDomIndex();
        let lightTypeDomIndex = ComponentDomTool.getLightTypeDomIndex();

        OperateComponentEventTool.addComponentIntoCurrentGameObject(
          boxComponentCount,
          renderingCategoryDomIndex,
          lightTypeDomIndex,
        );

        BuildComponentTool.buildInspectorComponent(
          TestTool.buildEmptyAppState(),
          InspectorTool.buildFakeAllShowComponentConfig(),
        )
        |> ReactTestTool.createSnapshotAndMatch;
      });
    });
    describe("test camera add component workflow", () => {
      beforeEach(() => {
        MainEditorSceneTool.createDefaultScene(
          sandbox,
          MainEditorSceneTool.setCameraTobeCurrentSceneTreeNode,
        );

        CurrentSelectSourceEditorService.setCurrentSelectSource(
          EditorType.SceneTree,
        )
        |> StateLogicService.getAndSetEditorState;
      });
      test("click the add component button, show addableComponent list", () => {
        let component =
          BuildComponentTool.buildInspectorComponent(
            TestTool.buildEmptyAppState(),
            InspectorTool.buildFakeAllShowComponentConfig(),
          );
        let cameraComponentCount = ComponentDomTool.getCameraComponentCount();

        BaseEventTool.triggerComponentEvent(
          component,
          OperateComponentEventTool.triggerClickShowComponentList(
            cameraComponentCount,
          ),
        );

        component |> ReactTestTool.createSnapshotAndMatch;
      });
      test("click arcballCameraController, add to inspector", () => {
        let cameraComponentCount = ComponentDomTool.getCameraComponentCount();
        let cameraCategoryDomIndex =
          ComponentDomTool.getCameraCategoryDomIndex();
        let arcballCameraTypeDomIndex =
          ComponentDomTool.getArcballCameraControllerTypeDomIndex();

        OperateComponentEventTool.addComponentIntoCurrentGameObject(
          cameraComponentCount,
          cameraCategoryDomIndex,
          arcballCameraTypeDomIndex,
        );

        BuildComponentTool.buildInspectorComponent(
          TestTool.buildEmptyAppState(),
          InspectorTool.buildFakeAllShowComponentConfig(),
        )
        |> ReactTestTool.createSnapshotAndMatch;
      });
    });
    describe("deal with specific case", () => {
      beforeEach(() =>
        MainEditorSceneTool.createDefaultScene(
          sandbox,
          MainEditorSceneTool.setFirstBoxTobeCurrentSceneTreeNode,
        )
      );
      test("if component type is error, should throw error", () =>
        expect(() =>
          InspectorComponentUtils.addComponentByType(
            "MeshTest",
            GameObjectTool.unsafeGetCurrentSceneTreeNode(),
          )
          |> StateLogicService.getEngineStateToGetData
        )
        |> toThrowMessage("the type:MeshTest is not find")
      );
    });
  });