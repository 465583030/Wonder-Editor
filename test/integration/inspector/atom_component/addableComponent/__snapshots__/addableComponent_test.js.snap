// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddableComponent test camera add component workflow click arcballCamera, add to inspector 1`] = `
<article
  className="wonder-inspector-component"
>
  <article
    className="wonder-inspector-sceneTree"
  >
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          Name
        </div>
      </div>
      <div>
        <article
          className="wonder-string-input"
        >
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="camera"
          />
        </article>
      </div>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          transform
        </div>
      </div>
      <article
        className="wonder-inspector-transform"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            X : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Y : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Z : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="40"
          />
        </article>
      </article>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          basicCameraView
        </div>
        <span
          className="header-close"
        >
          x
        </span>
      </div>
      <div>
        simulate basic camera view
      </div>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          perspectiveCameraProjection
        </div>
        <span
          className="header-close"
        >
          x
        </span>
      </div>
      <div>
        simulate perspective camera view
      </div>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          arcballCamera
        </div>
        <span
          className="header-close"
        >
          x
        </span>
      </div>
      <article
        className="wonder-inspector-transform"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            distance : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="10"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            min distance : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0.05"
          />
        </article>
      </article>
    </article>
    <article
      className="wonder-addable-component"
    >
      <button
        disabled={true}
        onClick={[Function]}
      >
        add component
      </button>
    </article>
  </article>
</article>
`;

exports[`AddableComponent test camera add component workflow click the add component button, show addableComponent list 1`] = `
<article
  className="wonder-inspector-component"
>
  <article
    className="wonder-inspector-sceneTree"
  >
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          Name
        </div>
      </div>
      <div>
        <article
          className="wonder-string-input"
        >
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="camera"
          />
        </article>
      </div>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          transform
        </div>
      </div>
      <article
        className="wonder-inspector-transform"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            X : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Y : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Z : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="40"
          />
        </article>
      </article>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          basicCameraView
        </div>
        <span
          className="header-close"
        >
          x
        </span>
      </div>
      <div>
        simulate basic camera view
      </div>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          perspectiveCameraProjection
        </div>
        <span
          className="header-close"
        >
          x
        </span>
      </div>
      <div>
        simulate perspective camera view
      </div>
    </article>
    <article
      className="wonder-addable-component"
    >
      <button
        disabled={false}
        onClick={[Function]}
      >
        add component
      </button>
      <div
        onClick={[Function]}
      >
        arcballCamera
      </div>
    </article>
  </article>
</article>
`;
